src/optimization/ControlFlow.cpp

```diff
@@ -1199,87 +1355,207 @@ bool optimizations::removeConstantLoadInLoops(const Module& module, Method& meth
-                                auto targetInst = targetBlock->key->walkEnd();
-                                targetInst.emplace(it.release());
-                            }
-                            else
-                            {
-                                CPPLOG_LAZY(logging::Level::DEBUG,
-                                    log << "Create a new basic block before the root of inclusion tree"
-                                        << logging::endl);
+                            targetCFGNode = *bb;
+                            break;
+                        }
+                    }
+                }
+            }
 
-                                auto headBlock = method.begin();
+            auto targetBlock = targetCFGNode != nullptr ? targetCFGNode->key : insertedBlock;
 
-                                insertedBlock = &method.createAndInsertNewBlock(
-                                    method.begin(), "%createdByRemoveConstantLoadInLoops");
-                                insertedBlock->walkEnd().emplace(it.release());
+            if(targetBlock != nullptr)
+            {
+                // insert before 'br' operation
+                auto& targetInst = targetBlock->walkEnd().previousInBlock();
+                for(auto it : insts->second)
+                {
+                    auto inst = it.get();
+                    if(!it.has() || processedInsts.find(inst) != processedInsts.end())
+                        continue;
+                    processedInsts.insert(inst);
 
-                                if(headBlock->getLabel()->getLabel()->name == BasicBlock::DEFAULT_BLOCK)
-                                {
-                                    // swap labels because DEFAULT_BLOCK is treated as head block.
-                                    headBlock->getLabel()->getLabel()->name.swap(
-                                        insertedBlock->getLabel()->getLabel()->name);
-                                }
-                            }
-                        }
-                        it.erase();
-                        hasChanged = true;
-                    }
+                    targetInst.emplace(it.release());
+                    it.erase();
+                }
+            }
+            else
+            {
+                logging::debug() << "Create a new basic block before the target block" << logging::endl;
+
+                auto headBlock = method.begin();
+
+                insertedBlock = &method.createAndInsertNewBlock(method.begin(), "%createdByRemoveConstantLoadInLoops");
+                for(auto it : insts->second)
+                {
+                    auto inst = it.get();
+                    if(!it.has() || processedInsts.find(inst) != processedInsts.end())
+                        continue;
+                    processedInsts.insert(inst);
+
+                    insertedBlock->walkEnd().emplace(it.release());
+                    it.erase();
+                }
```

Comment:
In the `--test-optimizations` tests, there is a case where the above `if`-clause accesses an instruction walker already freed in the `else`-clause, since the `insts->second` vector is traversed twice (first time with `else`-clause, second time with `if`-clause).
Anyway, inserting a `insts->second.clear();` here (after the closing bracket of the `else`-clause) fixes this crash. I'm not sure whether it is the correct thing to do though...